trezor/trezor-suite
# feat: add pre-pending txs for btc-like coins

# (pre)pending transactions

### General notes
- [x] when sending to a discovered account, we may display "prepending" transaction right away and we don't need to wait for blockchain notification. This is particulary needed for coinjoin accounts but also useful for all other accounts. 
- prepending tx should disappear when:
   - [x] fresh tx with the same txid arrives from backend
   - [x] fresh txs arrive from backend and current blockheight is 10 blocks higher then the blockeight at the moment when the prepending tx was created
- [x] prepending tx should not allow RBF

## (pre)pending 
The checkboxes below should represent the optimal state (probably?). Some of them are checked, some of them not. I'd say it is good enough for now and the rest could be solved in a followup

### Sending account (Btc-like non CJ) 
- [x]  balance decreases
- [x] utxo removed/updated
- [ ] change address -> next change address
### Receiving account (Btc-like non CJ)
- [x] balance increases
- [ ] utxo added
- [ ] unused address -> used address
### Sending account (Coinjoin)
- [x] balance decreases
- [ ] utxo removed/updated
- [ ] change address -> next change address
### Receiving account (Coinjoin)
- [x] balance increases
- [ ] utxo added
- [ ] unused address -> used address


## QA tips: 
- try slower your connection, use airplane mode after broadcasting tx
- check prepending/pending for btc like
- check possible regressions with cardano
- added a new [test](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/3832330013/artifacts/file/packages/suite-web/e2e/videos/pending-transactions.test.ts.mp4)
